modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java

```diff
@@ -211,4 +212,40 @@ void failAsClosedChannel() {
             buf.release();
         }
     }
+
+    private static final class ByteBufRefCounted implements RefCounted {
+
+        private final ByteBuf buffer;
+
+        ByteBufRefCounted(ByteBuf buffer) {
+            this.buffer = buffer;
+        }
+
+        @Override
+        public void incRef() {
+            buffer.retain();
+        }
+
+        @Override
+        public boolean tryIncRef() {
+            try {
+                buffer.retain();
+            } catch (RuntimeException e) {
+                assert refCount() == 0 : "should only fail if fully released but ref count was [" + refCount() + "]";
```

Comment:
Not the most elegant solution, but Netty ref counting simply doesn't have this functionality. Also, doesn't really matter as in practice the `tryIncRef` call seems not be used on the network buffers so there's performance relevant impact here.

Reply:
Perhaps check if `buffer.refCnt() == 0` first? We still need to catch an exception here but it'd be much rarer.

Reply:
++ makes sense